### PR TITLE
FPAY-14 test automation

### PR DESCRIPTION
**Pull request by satya-sanikommu**

### CODE REVIEW

File: src/Tests/Helpers/FundingPeriodAssertions.cs [hunk 1]
```diff
@@ -19,9 +19,9 @@ public class FundingPeriodAssertions : GenericCollectionAssertions<FundingPeriod
         {
             Execute.Assertion
                 .Given(() => subjArray[i])
-                .ForCondition(v => v.AgreedPrice == targetValue)
+                .ForCondition(v => v.AgreedPrice*0.80m == targetValue)
                 .FailWith(
-                    $"Expected value {subjArray[i].AgreedPrice}] in Period[{i}] should be {targetValue}");
+                    $"Expected value {subjArray[i].AgreedPrice} in Period[{i+1}] should be {targetValue}");
```
Comment (tom-gough) [inline src/Tests/Helpers/FundingPeriodAssertions.cs:24]: What is happening here?

File: src/Tests/local.settings.json [hunk 2]
```diff
@@ -1,8 +1,8 @@
 ﻿{
   "AzureWebJobsStorage": "UseDevelopmentStorage=true",
   "FUNCTIONS_WORKER_RUNTIME": "dotnet",
-  //"NServiceBusConnectionString": "at-shared-ns.servicebus.windows.net",
-  "NServiceBusConnectionString": "UseLearningEndpoint",
+  "NServiceBusConnectionString": "das-at-shared-ns.servicebus.windows.net",
+  //"NServiceBusConnectionString": "UseLearningEndpoint",
   "NServiceBusLicense": "",
   "LearningTransportStorageDirectory": "C:\\git\\das\\das-funding-apprenticeship-earnings\\src\\.learningtransport",
 }
```
Comment (tom-gough) [inline src/Tests/local.settings.json:8]: local.settings.json should not be checked in

Comment (tom-gough) [reply src/Tests/local.settings.json:8]: Also I think .learningtransport should be included in the .gitignore - not really a specific issue to this PR but we should fix it

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs [hunk 3]
```diff
@@ -1,60 +1,79 @@
...
     {
         _context = context;
+        _messageHelper = new ServiceBusMessageHelper(_context);
     }
 
-    [Given(@"An apprenticeship is created with (.*), (.*), (.*)")]
-    public async Task GivenAnApprenticeshipIsCreatedWith(decimal agreedPrice, DateTime actualStartDate,
-        DateTime plannedEndDate)
+    [Given(@"an apprenticeship has a start date of (.*), a planned end date of (.*), and an agreed price of �(.*)")]
+    public void AnApprenticeshipIsCreatedWith(DateTime start_date, DateTime planned_end_date, decimal agreed_price)
```
Comment (tom-gough) [inline src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:20]: variable names should be camel case

File: src/Tests/Helpers/ServiceBusMessageHelper.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        private ApprenticeshipCreatedEvent _apprenticeshipCreatedEvent;
+        private EarningsGeneratedEvent _earnings;
+        private FundingPeriod _funding;
+
+        public ServiceBusMessageHelper(ScenarioContext context)
+        {
+            _context = context;
+        }
+
+        public void CreateAnApprenticeshipMessage(DateTime actualStartDate, DateTime plannedEndDate, decimal agreedPrice)
```
Comment (tom-gough) [inline src/Tests/Helpers/ServiceBusMessageHelper.cs:15]: This method name hides the fact it's overwriting whatever is in _apprenticeshipCreatedEvent. Maybe drop the "An" from this and the publish method?

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs [hunk 5]
```diff
@@ -1,60 +1,79 @@
...
-        firstFundingPeriod.DeliveryPeriods.ForEach(dp => dp.LearningAmount.Should().Be(learningAmount));
+    [Then(@"the instalment amount must be calculated by dividing the total on-program amount equally into the number of planned months (.*)")]
+    public void VerifyInstallmentAmountIsCalculatedEquallyIntoAllEarningMonths(decimal installmentAmount)
+    {
+        _context.Get<FundingPeriod>().DeliveryPeriods.ForEach(dp => dp.LearningAmount.Should().Be(installmentAmount));
+    }
 
-        firstFundingPeriod.DeliveryPeriods.ShouldHaveCorrectFundingPeriods(numberOfInstalments, firstDeliveryPeriod, firstDeliveryAcademicYear);
-        firstFundingPeriod.DeliveryPeriods.ShouldHaveCorrectFundingCalendarMonths(numberOfInstalments, firstCalendarPeriodMonth, firstCalendarPeriodYear);
+    [Then(@"Earnings generated for each month starting from the first delivery period R(.*)-(.*) and first calendar period (.*)/(.*)")]
```
Comment (tom-gough) [inline src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:50]: Do delivery periods start with R? (Collection periods do, but I thought delivery periods did not)

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs [hunk 6]
```diff
@@ -1,60 +1,79 @@
...
+    {
+        var deliveryPeriods = _context.Get<FundingPeriod>().DeliveryPeriods;
+
+        int numberOfInstallments = deliveryPeriods.Count;
+
+        deliveryPeriods.ShouldHaveCorrectFundingPeriods(numberOfInstallments, firstDeliveryPeriodMonth, firstDeliveryPeriodYear);
+        deliveryPeriods.ShouldHaveCorrectFundingCalendarMonths(numberOfInstallments, firstCalendarPeriodMonth, firstCalendarPeriodYear);
+    }
+
+    //The following 2 methods are for the next set of tickets
```
Comment (tom-gough) [inline src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:61]: If these aren't needed for this ticket can we stick them on a branch for the next ticket instead?

**Review by tom-gough — CHANGES_REQUESTED** [review]

File: src/Tests/Helpers/FundingPeriodAssertions.cs (hunk 1 above)
Comment (MichaelYoung1981) [reply src/Tests/Helpers/FundingPeriodAssertions.cs:24]: I agree with @tom-gough although I'd go further and say we shouldn't have logic in test code.

File: src/Tests/Helpers/ServiceBusMessageHelper.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            var fixture = new Fixture();
+            _apprenticeshipCreatedEvent = fixture.Build<ApprenticeshipCreatedEvent>()
```
Comment (MichaelYoung1981) [inline src/Tests/Helpers/ServiceBusMessageHelper.cs:18]: I don't think this is the right event, we need to be using the one from approvals.

File: src/Tests/Features/CalculateEarnings.feature [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+So they feed into payments calculation I get paid
+
+@regression
+Scenario: Earnings Generation for an approved apprenticeship
+	Given an apprenticeship has a start date of <start_date>, a planned end date of <planned_end_date>, and an agreed price of £<agreed_price>
+	When the apprenticeship commitment is approved
+	Then the total on-program payment amount must be calculated as 80% of the agreed price £<adjusted_price>
+	And the planned number of months must be the number of months from the start date to the planned end date <planned_number_of_months>
+	And the instalment amount must be calculated by dividing the total on-program amount equally into the number of planned months <instalment_amount>
+	And Earnings generated for each month starting from the first delivery period <first_delivery_period> and first calendar period <first_calendar_period>
```
Comment (MichaelYoung1981) [inline src/Tests/Features/CalculateEarnings.feature:14]: This step should match the step in the user story.

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs [hunk 9]
```diff
@@ -1,60 +1,79 @@
...
+    private decimal CalculateOnProgramPaymentBasedOnAgreedPriceAndFundingBand(decimal agreed_price, decimal fundingband_value)
     {
-        _earnings.FundingPeriods.First().Uln.Should().Be(Convert.ToInt64(_apprenticeshipCreatedEvent.Uln));
-        _earnings.FundingPeriods.First().EmployerId.Should().Be(_apprenticeshipCreatedEvent.EmployerAccountId);
-        _earnings.FundingPeriods.First().ProviderId.Should().Be(_apprenticeshipCreatedEvent.UKPRN);
-        _earnings.FundingPeriods.First().TransferSenderEmployerId.Should().Be(_apprenticeshipCreatedEvent.FundingEmployerAccountId);
-        _earnings.FundingPeriods.First().StartDate.Should().Be(_apprenticeshipCreatedEvent.ActualStartDate);
-        _earnings.FundingPeriods.First().TrainingCode.Should().Be(_apprenticeshipCreatedEvent.TrainingCode);
-        _earnings.FundingPeriods.First().EmployerType.Should().Be(_apprenticeshipCreatedEvent.FundingType == FundingType.NonLevy ? EmployerType.NonLevy : EmployerType.Levy);
+        return agreed_price * 0.80m;
```
Comment (MichaelYoung1981) [inline src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:77]: I'd much rather we didn't calculate values in tests, they should be part of the feature.

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs (hunk 3 above)
Comment (satya-sanikommu) [reply src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:20]: Done

File: src/Tests/Helpers/ServiceBusMessageHelper.cs (hunk 4 above)
Comment (satya-sanikommu) [reply src/Tests/Helpers/ServiceBusMessageHelper.cs:15]: Updated

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs (hunk 6 above)
Comment (satya-sanikommu) [reply src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:61]: Removed from this

File: src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs (hunk 9 above)
Comment (satya-sanikommu) [reply src/Tests/StepDefinitions/CalculateEarningsForLearningPaymentsStepDefinitions.cs:77]: This is just a placeholder one to implement for the following tickets. But this is removed now

File: src/Tests/Features/CalculateEarnings.feature (hunk 8 above)
Comment (satya-sanikommu) [reply src/Tests/Features/CalculateEarnings.feature:13]: Discussed this with the devs and agreed to update the spec in the story as this is more appropriate but achieves the same result

File: src/Tests/Helpers/FundingPeriodAssertions.cs (hunk 1 above)
Comment (satya-sanikommu) [reply src/Tests/Helpers/FundingPeriodAssertions.cs:24]: This is no longer used.

Comment (tom-gough) [reply src/Tests/Helpers/FundingPeriodAssertions.cs:24]: should we remove it then?

**Review by tom-gough — APPROVED** [review]

File: src/Tests/Helpers/FundingPeriodAssertions.cs (hunk 1 above)
Comment (satya-sanikommu) [reply src/Tests/Helpers/FundingPeriodAssertions.cs:24]: Removed

**Review by ruslanss — APPROVED** [review]

**Review by EwanNoble — APPROVED** [review]

**Pull request merged**